File: Daily/Matrix_Similarities_After_Cyclical_Shift/matrix__similarities__after__cyclical__shift.py
from typing import Optional, List
import heapq
from collections import defaultdict


class Solution:
    def areSimilar(self, mat: List[List[int]], k: int) -> bool:
        n = len(mat[0])
        if n <= 1:
            return True

        k = k % n

        for row in mat:
            if len(set(row)) > 1:
                for i in range(n):
                    if row[i] != row[(k + i) % n]:
                        return False
        return True


if __name__ == "__main__":
    sol = Solution()

    mat = [[1, 2, 1, 2], [5, 5, 5, 5], [6, 3, 6, 3]]
    k = 2
    print(sol.areSimilar(mat, k))

    mat = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
    k = 4
    print(sol.areSimilar(mat, k))

    print("Running Solution...")

File: Daily/Matrix_Similarities_After_Cyclical_Shift/Matrix_Similarities_After_Cyclical_Shift.cpp
// Matrix_Similarities_After_Cyclical_Shift.cpp
//
// Compile locally with:
// clang++ Matrix_Similarities_After_Cyclical_Shift.cpp -std=c++17 -Wall -Wextra -O2 -o run && ./run
//
// NOTE:
// - Remove main() before submitting to LeetCode.
// - Keep only the class Solution definition.
//

#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    bool areSimilar(vector<vector<int>> &mat, int k)
    {
        int n = mat[0].size();

        for (vector<int> row : mat)
        {
            for (int idx = 0; idx < n; idx++)
            {
                if (row[idx] != row[(k + idx) % n])
                {
                    return false;
                }
            }
        }
        return true;
    }
};

int main()
{
    cout << "Running Matrix_Similarities_After_Cyclical_Shift.cpp..." << endl;

    Solution sol;

    // TODO:
    // Add local test calls here

    return 0;
}
